cmpserver/plugin/plugin.go

```diff
@@ -143,24 +143,49 @@ func environ(envVars []*apiclient.EnvEntry) []string {
 	return environ
 }
 
+// getTempDirMustCleanup creates a temporary directory and returns a cleanup function. The cleanup function panics if
+// cleanup fails. Use this function when failing to clean up the temporary directory is a security risk.
+func getTempDirMustCleanup(baseDir string) (workDir string, cleanup func(), err error) {
+	workDir, err = files.CreateTempDir(baseDir)
+	if err != nil {
+		return "", nil, fmt.Errorf("error creating temp dir: %w", err)
+	}
+	cleanup = func() {
+		if err := os.RemoveAll(workDir); err != nil {
+			// we panic here as the workDir may contain sensitive information
+			panic(fmt.Sprintf("error removing plugin workdir: %s", err))
```

Comment:
I think the panic would make sense if we couple it with a panic on container startup if we fail to clean up the dir.

I'd like to strongly guarantee cleanup, since it's one of the main ways we mitigate traversal bugs in the sidecar.

Reply:
My argument is that if you really want to panic, then that should be done in a upper layer in the application. Ideally in the main function. Having panic in inner functions this way is not ideal as different consumers will likely handle the error scenario differently. This can be simply accomplished by having something like:
`cleanup = func() error`
And then the caller is the one responsible of handling that error with a panic or not.

Reply:
Sure... come to think of it, I don't even think the panic would restart the container. I bet the server library catches it.

Reply:
We do have an interceptor to recover from panic in the grpc server:
https://github.com/argoproj/argo-cd/blob/master/cmpserver/server.go#L52

This doesn't mean however that we should be using panic as a common practice in the codebase. I agree that we should strongly guarantee cleanup but it is unclear to me how this is achieved by the current implementation. Maybe we can achieve that without relying on panics?